src/transaction/transaction.rs

```diff
@@ -542,62 +597,95 @@ struct TwoPhaseCommitter {
+
+        let min_commit_ts = self.prewrite().await?;
+
+        if self.style.try_one_pc {
+            return Ok(min_commit_ts);
+        }
+
+        let commit_ts = if self.style.async_commit {
+            assert_ne!(min_commit_ts, 0);
+            min_commit_ts
+        } else {
+            match self.commit_primary().await {
+                Ok(commit_ts) => commit_ts,
+                Err(e) => {
+                    return if self.undetermined {
+                        Err(Error::undetermined_error(e))
+                    } else {
+                        Err(e)
+                    };
                 }
             }
-        }
+        };
+        self.bg_worker
+            .clone()
+            .spawn_ok(self.commit_secondary(commit_ts).map(|res| {
+                if let Err(e) = res {
+                    warn!("Failed to commit secondary keys: {}", e);
+                }
+            }));
+        Ok(commit_ts)
     }
 
-    async fn prewrite(&mut self) -> Result<()> {
+    async fn prewrite(&mut self) -> Result<u64> {
         let primary_lock = self.mutations[0].key.clone().into();
         // TODO: calculate TTL for big transactions
         let lock_ttl = DEFAULT_LOCK_TTL;
-        if self.for_update_ts > 0 {
-            new_pessimistic_prewrite_request(
+        let mut request = match self.style.kind {
+            TransactionKind::Optimistic => new_prewrite_request(
                 self.mutations.clone(),
                 primary_lock,
                 self.start_version,
                 lock_ttl,
-                self.for_update_ts,
-            )
-            .execute(self.rpc.clone(), PESSIMISTIC_BACKOFF)
-            .await
-        } else {
-            new_prewrite_request(
+            ),
+            TransactionKind::Pessimistic(for_update_ts) => new_pessimistic_prewrite_request(
                 self.mutations.clone(),
                 primary_lock,
                 self.start_version,
                 lock_ttl,
-            )
-            .execute(self.rpc.clone(), OPTIMISTIC_BACKOFF)
-            .await
+                for_update_ts,
+            ),
+        };
+
+        request.use_async_commit = self.style.async_commit;
+        request.try_one_pc = self.style.try_one_pc;
+        request.secondaries = self.mutations[1..].iter().map(|m| m.key.clone()).collect();
+        // FIXME set max_commit_ts and min_commit_ts
+
+        let response = match self.style.kind {
+            TransactionKind::Optimistic => {
+                request
+                    .execute(self.rpc.clone(), OPTIMISTIC_BACKOFF)
+                    .await?
+            }
+            TransactionKind::Pessimistic(_) => {
+                request
+                    .execute(self.rpc.clone(), PESSIMISTIC_BACKOFF)
+                    .await?
+            }
+        };
+
+        if self.style.try_one_pc {
+            if response.one_pc_commit_ts == 0 {
+                return Err(ErrorKind::OnePcFailure.into());
```

Comment:
Will `OnePcFailure` fallback to 2PC (in next PR)?

Reply:
Yes, I think so, at least as an option